File: DSA + C++/35 Recursion/Binary_search.cpp
#include<bits/stdc++.h>
using namespace std;

int binary_search(int key,int* arr,int s,int e){
    int mid=(e-s)/2+s;
    
    if(arr[mid]==key)
    return mid;
    
    if (s==e)
    return -1;

    if(arr[mid]>key)
    binary_search(key,arr,s,mid-1);
    else
    binary_search(key,arr,mid+1,e);


}

int main()
{
int n;
cin>>n;
int arr[n];
for (int i = 0; i < n; i++)
{
    cin>>arr[i];
}
int key;
cin>>key;
int s=0;
int e=n-1;
cout<<binary_search(key,arr,s,e);

return 0;
}

File: PY/DSA/Binary_search.py

def binary_search(list,key):
    s=0
    e=len(list)-1
    while s<=e:
        mid=int((s+e)/2)
        if(list[mid]==key):
            return mid
        if(list[mid]>key):
            e=mid-1
        else:
            s=mid+1
    return None


my_list=[3,5,6,8,10]

print("3 is at index: "+str(binary_search(my_list,9)))